### DISCUSSION TITLE
When to use embedded smart contracts?

### DISCUSSION THREAD

**Discussion by georgezgeorgez** [Big-Inches-Club-House/bich#5] [General]
This is probably one of the bigger conversations that we need to have in terms of network design philosophy.

I had previously written about some of the differences between embedded smart contracts and VM smart contracts here:

https://github.com/Big-Inches-Club-House/bich/blob/master/newsletter/3.md#smart-contracts

After implementing a prototype for an htlc embedded contract, I have some more experience on the matter.

In my last writing I called out:

1. Access vs Isolation
2. Upgradability vs Immutability
3. Performance

For 1) I had mentioned that embedded smart contracts potentially had more access to critical network data while a properly implemented VM would provide proper isolation.

While this is still true, I find that the design of the embedded smart contract system provides a decent amount of safety. It seems like one would have to explicitly try to break the scope of data in order to create side effects outside the scope of the contract.

2) remains an important consideration

3) again hard to say without real profiling

But I think the main driver of whether or not something should be an embedded smart contract is something I didn't really mention last time.

Namely,
1. Is the functionality core to the protocol?
2. If not, is the functionality expected to follow an established standard?
3. If not, do we want to provide an opinionated standard?

And finally, as protocol developers, is it something we want to include in protocol governance?

It is likely that for anything related to emissions, we will want it defined explicitly as part of the node code.

In Ethereum, many smart contracts are written to work with ERC-20 tokens. So often ETH must first be wrapped to utilize these contracts. The Zenon founding developers decided to implement a token standard (ZTS) to avoid this, and to define ZNN and QSR as ZTS tokens. In doing so they provided an opinionated standard and implementation for tokens.

ERC-20 is just a interface and people who remember the wZNN LP hack can appreciate a standard implementation for tokens. There is no need to audit every single ZTS token to ensure there are no backdoors or vulnerabilities.

On the other hand, the standard is opinionated and dictates that every ZTS token needs a url. And through my experimentation with BTC709632 ZTS proto NFTs (https://github.com/Big-Inches-Club-House/bich/blob/master/newsletter/3.md#nfts), I discovered to my annoyance the url wouldn't accept native ipfs addresses, and had to find a work around.

Here the desire to define the network native currencies ZNN and QSR through a protocol level token standard necessitated the creation of the ZTS embedded smart contract.

And I believe that most of the functionality needed for the core protocol incentives (emissions) already has embedded smart contracts created or stubbed out with a placeholder (eg liquidity).

So we'll move on to consider functionality that is not strictly necessary for the protocol incentives but is relatively well defined.

One example of this might be Constant Product Liquidity Pools / Automated Market Makers. There is a mathematical formula. 

While the design of a generic VM runtime entails quite a lot of design decisions, an AMM embedded smart contract would be relatively simple to design and implement, without needing to be too opinionated. And it would provide immediate value to the network.

One issue facing the community now is the illiquidity of QSR. Community members with the means and motivation to run infrastructure cannot acquire QSR quickly without OTC. 

In addition, people with testnet QSR airdrops have no way to meaningfully exchange it and participate in the protocol.

A QSR/ZNN pool created through an embedded amm smart contract could quickly solve this while allowing for the creation of any other ZTS pair pools.

I think there are arguments to be made for incentivizing a QSR/ZNN pool through a portion of liquidity emissions. Implementing the pool through an embedded smart contract could make calculations of liquidity rewards much more streamlined. Rather than needing to rely on a layer of indirection, through oracles or a interface to be implemented by a vm contract.

Although I have a working prototype for htlcs through embedded smart contracts, I'm not quite sure the functionality is deserving of one. But it is something without too many opinionated design decisions and may be useful for power user liquidity while we work out the design for a VM.

One thing that the community has been wanting but perhaps has been disappointed about is NFTs.

As explained in my explanation of the BTC709632 NFTs, ERC-721 and ERC-1155 provide token standards and OpenSea provides a set of metadata standards. However I view these as pretty opinionated and not as indisputable natural standards.

NFTs don't seem critical to protocol incentives/emissions. Is it worth creating an embedded smart contract for NFTs that implement an opinionated standard?

By providing an opinionated implementation, would we stifle potential innovation?

So I'm not really sure what rule of thumb we should follow when it comes to providing embedded smart contracts that don't implement core incentives.

While we don't have a VM defined, it is also more tempting to try and add some quick functionality before needing to fully consider a VM.

How important are these timing considerations? Should we focus less on specific embedded smart contracts and more on knocking the VM out first? Or maybe prototypes of these functionalities through embedded smart contracts will help provide context and requirements for a VM runtime.

Community guidance and feedback on approach is needed to help guide our efforts.

**Comment by georgezgeorgez** [comment]
The community answer could be as simple as no more embedded smart contracts right now, focus on the VM since we need it no matter what. But I don't want to rule out that there could be many instances where it makes senses to provide an embedded sc.

**Reply by 0x3639** [reply]
I've read these posts very carefully to be sure I understand everything. 

1) Do you think we should wait to learn more about Phase 1 before making a decision.  Sounds like you believe the VM will not be a part of Phase I which is a reasonable assumption.

2) Providing QSR / ZNN liquidity is important for sure.  I think that should be a top priority so if getting there fast involves an embedded SC  I think we should strongly consider that option.
- Can your HTLC code written be used for that?
- Should we consider an order book, rather than AMM for the liquidity solution?  

3) Would hate to lock into a rigid NFT standard now before understanding Phase I and more about VMs.

In general I think we should work towards a liquidity solution and then a VM implementation.  Even if the embedded SC code for an AMM / Order book is not "worthy" of one, the community could benefit from it and attract more pillars.

**Reply by georgezgeorgez** [reply]
My impression is that Phase 1 is more about finishing up the implementation of our key incentives and defining a scaling strategy. This is more about design philosophy and what guiding principles we should adhere to.

If Mr Kaine presents us with something concrete, that can factor in the discussion, but it's still a discussion to be had.

I agree that more liquidity is needed. Htlcs can help with QSR/ZNN but to be honest, as I mentioned in my post, using HTLCs directly will probably be more for power users. 

AMMs are much easier in comparison. 

I don't think an order book model would fit well.

**Comment by vilkris4** [comment]
Adding some reasonably unopinionated embedded contracts at this stage could be a relatively quick way to add more features to the network opening up more use cases for community members to build on. I do think that implementing some embedded contracts could give valuable insight into what should be considered when implementing a VM runtime.

I also don’t see why there couldn’t be competing implementations of these functionalities. One could be an embedded contract and the other could be a contract running on the VM once it's possible.

Any thoughts on how feasible it would be to later deprecate an embedded contract, if that would be needed for whatever reason?

**Reply by georgezgeorgez** [reply]
Yes I think it will be worthwhile to have some more prototype embedded smart contracts for the community to consider in any case. 

An AMM sounds like it could be a reasonable candidate.

There's no reason why someone wouldn't be able to develop an AMM on the VM even if there was a embedded contract for it. 

Changes including deprecation shouldn't be too hard. Might need to disable some functions. Eg only allow withdrawals. Or force an exit of all funds in it.

Maybe it's unnecessary, but if we (HyperCore) create an embedded smart contract, it feels like we are obligated to deliver a Syrius update as well so people can actually use it.

Whereas for embedded smart contracts, the writer of the smart contract would be responsible. 

Just some governance/maintenance concerns to keep in mind.

I've mentioned it before. But long term I would love for Syrius to have a plug-in architecture where people could download plugins for their favorite smart contracts widgets.

**Reply by vilkris4** [reply]
An AMM could definitely be a reasonable candidate and the benefit of it is obvious.

It's true that providing a user interface and a proper UX for using the embedded contracts is important. Doing the AMM contract and a UI for it on syrius seems like something that could be achievable right now.

**Comment by georgezgeorgez** [comment]
A contract can be thought of as a few parts
- An account which can hold native ZTS tokens
- State within it's own level db scope (usually to track owners of held tokens)
- And logic triggered by txs which can modify the state or make tx

As I'm thinking about the ability to upgrade and deprecate contracts, it's useful to think about what changes are required. 

Schema of the state? Logic?
Both?

Often for relational databases, schema changes are done through migration scripts. 

For schema changes to state, we could similarly use migration scripts. For embedded smart contracts, these scripts could be part of the node binary, or even reference data on chain. Technically all logic could be defined through on chain data. 

Referencing data on chain would make it harder to launch other networks. As the genesis config would not just have to specify a hash ID for a spork, but also code. 
But this would allow for other networks to only need to include the latest schema/state and logic in its genesis.

Keeping it all in the node code, we probably end up in a situation where every time we make protocol level changes, we expand the size of the code/binary. 

So I'm not sure which approach we want to take. 
More things in the node code, or increasingly change the protocol to rely on on-chain objects.  

For smart contracts, it could only reference on chain objects.

It could be possible to create a generic smart contract framework that also allowed for the concept of state migration scripts. And by a layer of indirection have mutable logic. What triggers a change is a different discussion.
